pkg/cmd/pulumi/import.go

```diff
@@ -0,0 +1,252 @@
+			"should be added to the Pulumi program. The resource is protected from deletion\n" +
+			"by default",
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			opts := display.Options{
+				Color: cmdutil.GetGlobalColorization(),
+			}
+
+			// Fetch the project.
+			project, _, err := readProject()
+			if err != nil {
+				return err
+			}
+
+			var programGenerator func(p *hcl2.Program) (map[string][]byte, hcl.Diagnostics, error)
+			switch project.Runtime.Name() {
+			case "dotnet":
+				programGenerator = dotnet.GenerateProgram
+			case "go":
+				programGenerator = gogen.GenerateProgram
+			case "nodejs":
+				programGenerator = nodejs.GenerateProgram
+			case "python":
+				programGenerator = python.GenerateProgram
+			default:
+				return fmt.Errorf("cannot generate resource definitions for %v", project.Runtime.Name())
+			}
+
+			// Fetch the current stack.
+			s, err := requireStack(stackName, false, opts, true /*setCurrent*/)
+			if err != nil {
+				return err
+			}
+			stackName := s.Ref().Name()
+
+			sm, err := getStackSecretsManager(s)
+			if err != nil {
+				return errors.Wrap(err, "getting secrets manager")
+			}
+
+			cfg, err := getStackConfiguration(s, sm)
+			if err != nil {
+				return errors.Wrap(err, "getting stack configuration")
+			}
+
+			// Export the stack's deployment.
+			deployment, err := s.ExportDeployment(context.Background())
+			if err != nil {
+				return err
+			}
+			snap, err := stack.DeserializeUntypedDeployment(deployment, stack.DefaultSecretsProvider)
+			if err != nil {
+				switch err {
+				case stack.ErrDeploymentSchemaVersionTooOld:
+					return fmt.Errorf("the stack '%s' is too old to be used by this version of the Pulumi CLI",
+						stackName)
+				case stack.ErrDeploymentSchemaVersionTooNew:
+					return fmt.Errorf("the stack '%s' is newer than what this version of the Pulumi CLI understands. "+
+						"Please update your version of the Pulumi CLI", stackName)
+				}
+
+				return errors.Wrap(err, "could not deserialize deployment")
+			}
+			stackResource, err := stack.GetRootStackResource(snap)
+			if err != nil {
+				return err
+			}
+
+			// Build a URN for the resource.
+			urn := resource.NewURN(stackName, project.Name, "", tokens.Type(args[0]), tokens.QName(args[1]))
+
+			// Create a plugin context and host.
+			target := &deploy.Target{
+				Name:      stackName,
+				Config:    cfg.Config,
+				Decrypter: cfg.Decrypter,
+				Snapshot:  snap,
+			}
+			cwd, err := os.Getwd()
+			if err != nil {
+				return err
+			}
+			sink := diag.DefaultSink(os.Stdout, os.Stderr, diag.FormatOptions{Pwd: cwd, Color: colors.Always})
+			ctx, err := plugin.NewContext(sink, sink, nil, target, cwd, nil, nil)
+			if err != nil {
+				return err
+			}
+			host := ctx.Host
+
+			// Load the default provider.
+			//
+			// TODO: allow a provider to be specified on the command line or pull from the snapshot, as we can't
+			// otherwise store to the state.
```

Comment:
This is important even without the bit about storing to the state. We really should generate the definition using the schema for the provider version that is present in the state file, if any.